packages/sqlite_async/lib/src/web/web_mutex.dart

```diff
@@ -17,12 +41,117 @@ class MutexImpl implements Mutex {
 
   @override
   Future<T> lock<T>(Future<T> Function() callback, {Duration? timeout}) {
-    // Note this lock is only valid in a single web tab
-    return m.protect(callback);
+    if ((_navigator as JSObject).hasProperty('locks'.toJS).toDart) {
+      return _webLock(callback, timeout: timeout);
+    } else {
+      return _fallbackLock(callback, timeout: timeout);
+    }
+  }
+
+  /// Locks the callback with a standard Mutex from the `mutex` package
+  Future<T> _fallbackLock<T>(Future<T> Function() callback,
+      {Duration? timeout}) {
+    final completer = Completer<T>();
+    // Need to implement timeout manually for this
+    bool isTimedOut = false;
+    bool lockObtained = false;
+    if (timeout != null) {
+      Future.delayed(timeout, () {
+        if (lockObtained == false) {
+          isTimedOut = true;
+          completer.completeError(
+              TimeoutException('Failed to acquire lock', timeout));
+        }
+      });
+    }
+
+    fallback.protect(() async {
+      try {
+        if (isTimedOut) {
+          // Don't actually run logic
+          return;
+        }
+        lockObtained = true;
+        final result = await callback();
+        completer.complete(result);
+      } catch (ex) {
+        completer.completeError(ex);
+      }
+    });
+
+    return completer.future;
+  }
+
+  /// Locks the callback with web Navigator locks
+  Future<T> _webLock<T>(Future<T> Function() callback,
+      {Duration? timeout}) async {
+    final lock = await _getWebLock(timeout);
+    try {
+      final result = await callback();
+      return result;
+    } finally {
+      lock.release();
+    }
+  }
+
+  /// Passing the Dart callback directly to the JS Navigator can cause some weird
+  /// context related bugs. Instead the JS lock callback will return a hold on the lock
+  /// which is represented as a [HeldLock]. This hold can be used when wrapping the Dart
+  /// callback to manage the JS lock.
+  /// This is inspired and adapted from https://github.com/simolus3/sqlite3.dart/blob/7bdca77afd7be7159dbef70fd1ac5aa4996211a9/sqlite3_web/lib/src/locks.dart#L6
+  Future<HeldLock> _getWebLock(Duration? timeout) {
+    final gotLock = Completer<HeldLock>.sync();
+    // Navigator locks can be timed out by using an AbortSignal
+    final controller = AbortController();
+
+    bool lockAcquired = false;
+    if (timeout != null) {
+      // Can't really abort the `delayed` call easily :(
```

Comment:
You can use a Timer instead of Future.delayed if you want to abort it

Reply:
This is much nicer. Implemented Timers now.